chain-impl-mockchain/src/leadership/genesis.rs

```diff
@@ -0,0 +1,1481 @@
+use crate::block::{BlockDate, Message, SignedBlock};
+use crate::certificate;
+use crate::key::PublicKey;
+use crate::leadership::bft::BftRoundRobinIndex;
+use crate::ledger::Ledger;
+use crate::setting::{self, Settings};
+use crate::transaction::Value;
+use crate::update::ValueDiff;
+
+use chain_addr::Kind;
+use chain_core::property::{self, Block, LeaderSelection, Update};
+
+use rand::{Rng, SeedableRng};
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug)]
+pub struct GenesisLeaderSelection {
+    ledger: Arc<RwLock<Ledger>>,
+    settings: Arc<RwLock<Settings>>,
+
+    bft_leaders: Vec<PublicKey>,
+
+    stake_keys: HashMap<PublicKey, StakeKeyInfo>,
+
+    stake_pools: HashMap<PublicKey, StakePoolInfo>,
```

Comment:
the stake keys and the stake pools could live within a same wrapping type. Especially since their is a link between both. It would make the rest of the code easier to read too. Also, please, utilise different type for the stake_keys and stake_pools. `PublicKey` is a generic type that is being overloaded by a lot of different meaning.

Reply:
Done, there now are types `StakeKeyId`, `StakePoolId`, and `LeaderId`.